src/test/java/com/wordCount/WordCounterIntegrationTests.java

```diff
@@ -0,0 +1,53 @@
+package com.wordCount;
+
+import com.wordCount.mocks.ConsoleTextReaderStub;
+import com.wordCount.mocks.ConsoleWriterSpy;
+import com.wordCount.paramsSource.TestParamValuesSource;
+import com.wordcount.controllers.WordCounterController;
+import com.wordcount.writers.AnswerWriter;
+import com.wordcount.writers.impl.AnswerWriterImpl;
+import org.junit.jupiter.api.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class WordCounterIntegrationTests {
+
+    @Test
+    public void word_counting_in_a_text_made_of_words_only_is_correct() {
+
+        ConsoleTextReaderStub readerStub = new ConsoleTextReaderStub();
+        ConsoleWriterSpy consoleWriterSpy = new ConsoleWriterSpy();
+        AnswerWriter answerWriter = new AnswerWriterImpl(consoleWriterSpy);
+
+        HashMap<String, Integer> params = TestParamValuesSource.getWordsOnlyParamValues();
+
+        for (Map.Entry<String, Integer> param : params.entrySet()) {
+            readerStub.setup(param.getKey());
+
+            WordCounterController sut = new WordCounterController(readerStub, answerWriter);
+            sut.countWords();
+
+            consoleWriterSpy.shouldWriteText(String.format("Number of words: %d", param.getValue()));
+        }
+    }
+
+    @Test
+    public void word_counting_in_a_text_made_not_of_words_only_is_correct() {
+
+        ConsoleTextReaderStub readerStub = new ConsoleTextReaderStub();
+        ConsoleWriterSpy consoleWriterSpy = new ConsoleWriterSpy();
+        AnswerWriter answerWriter = new AnswerWriterImpl(consoleWriterSpy);
+
+        HashMap<String, Integer> params = TestParamValuesSource.getNonWordsParamValues();
+
+        for (Map.Entry<String, Integer> param : params.entrySet()) {
+            readerStub.setup(param.getKey());
+
+            WordCounterController sut = new WordCounterController(readerStub, answerWriter);
+            sut.countWords();
+
+            consoleWriterSpy.shouldWriteText(String.format("Number of words: %d", param.getValue()));
+        }
+    }
```

Comment:
In general ok, but because of reusing test params you are duplicating the execution of number of test cases in this test class with the `WordCounterUnitTests`. Unit Tests are expected to have more test examples than integration tests. In integration tests you want to test only the `integration` of all collaborators. Therefore you could remove one of the test methods in this test.

With the help of a little improved English, static imports and renaming of variables you could improve readability quite a bit including a bit more conciseness.

```Java
private final ConsoleTextReaderStub reader = new ConsoleTextReaderStub();
private final ConsoleWriterSpy writer = new ConsoleWriterSpy();
private final AnswerWriter answerWriter = new AnswerWriterImpl(writer);
private final WordCounterController sut = new WordCounterController(reader, answerWriter);

@Test
public void count_words_in_given_text_which_contains_words() {
    paramsWithWords().forEach((userInputText, expectedWordCount) -> {
        reader.returns(userInputText);

        sut.countWords();

        writer.writes(format("Number of words: %d", expectedWordCount));
    });
}
```
I would actually be absolutely ok if you just test one single "integration" test case like:
```Java
@Test
public void count_words_in_given_text_which_contains_words() {
    reader.returns("Mary had a little lamb");
      
    sut.countWords();
        
    writer.writes(format("Number of words: %d", 5));
}
```
If your tests fail for whatever reason, your `ConsoleWriterSpy` yields following error message: 
```
org.opentest4j.AssertionFailedError: 
Expected :Number of words: 4
Actual   :Number of words: 5
```
Please try to optimise automated tests in general for **error detection**. Usually we come back to reading tests when 

1. we want to learn more about the system under test or 
2. if tests fail because of a bug. 



Then the error messages should be as expressive as possible to be able to detect errors as fast as possible. You could adapt your spy like this, to optimise for error detection
```Java
@Test
public void count_words_in_given_text_which_contains_words() {
    String userInputText = "Mary had a little lamb";
    reader.returns(userInputText);

    sut.countWords();

    writer
        .forInput(userInputText)                                  // <= New
        .writes(format("Number of words: %d", 4));
}
...
public class ConsoleWriterSpy implements ConsoleWriter {

    private String outputText;
    private String userInputText;

    @Override
    public void write(String outputText) {
        this.outputText = outputText;
    }

    public void writes(String expectedOutput) {
        assertEquals(expectedOutput, outputText, errorMessageFor(expectedOutput));
    }

    private String errorMessageFor(String expectedOutput) {
        return format("For given user input text: '%s' following output text: '%s' is expected",
                userInputText,
                expectedOutput);
    }

    public ConsoleWriterSpy forInput(String userInputText) {
        this.userInputText = userInputText;
        return this;
    }
}
``` 
which would yield the following error message:
````
org.opentest4j.AssertionFailedError: For given user input text: 'Mary had a little lamb' following output text: 'Number of words: 4' is expected ==> 
Expected :Number of words: 4
Actual   :Number of words: 5
´´´